### ISSUE TITLE
Pure C++ target (with optional Swift/Objective-C support)

### ISSUE THREAD

**Issue by soywiz** [jtransc/jtransc#51]
It was pretty easy to add the new pure-javascript target: https://github.com/jtransc/jtransc/tree/master/jtransc-gen-js

So investigate in future versions (after stabilizing stuff, probably after 1.0) to generate a C++ target with GC (and optionally ARC).

**Optional / Future / Optimization:**
For ARC there is already a `@JTranscWeak` annotation that could allow to specify weak stuff to avoid leaks.
Runtime should be updated to support this (annotating fields with the weak annotation) and code and libraries must be annotated too.
This could allow to use emscripten to generate ASM.js/WASM as commented here: https://github.com/jtransc/jtransc/issues/38

**Comment by soywiz** [comment]
This target could have GOTO support, and that means awesome performance out of the box even without relooping.
I have checked and both GCC and LLVM makes equally faster loop using a for, than emulating with ifs and gotos.

**Comment by dsp-omen** [comment]
I found this awesome project a couple of weeks ago.

I'm interested in compile some java code into C++ to be able to run it in a small microcontroller where a full JVM would be overkilling. Also I need zero pauses related to GC. And I found this.

I have seen that the project is not updated anymore, since July at least. Do you plan update it again? Are you still going to add C++ support and ARC support? Would you accept a contribution?

I have look through the code base, and even when it is not perfect, I think I can understand it pretty well.
My experience with Kotlin is limited, and never contributed to an opensource project before, but I have some experience related to compilers from the college. If you are not going to update this project anymore, would you consider a contribution related to this before 1.1?

Also, do you know any other project for compiling Java into C++ with ARC? I didn't find any other project that provides what actually need.

**Comment by soywiz** [comment]
Thanks :)

Sounds pretty interesting. I paused the project due to lack of time + motivation at this point. I plan to resume it eventually, but not right now.

Don't know about other options. I wanted to have javascript + C++. Right now you can generate C++ using haxe as intermediate, but as you said it uses a GC so it will generate pauses and it is not optimal. No other options did that.

Of course! Any contribution is really welcome. If you manage to get it working, I will definitively merge it even before 1.1. Thank you! Maybe you are interested too in this: https://github.com/jtransc/jtransc/issues/24

You can contact me if you have any issues or questions in the progress.

**Comment by dsp-omen** [comment]
Ok. I have started already. At this point I'm not used to git yet, usually I'm working offline with a subversion backup at a local-network computer. It is ok if for you if I send to you a zip file with the changes when done?

**Comment by soywiz** [comment]
Well, usually it is better to do a [Pull Request](https://help.github.com/articles/about-pull-requests/) instead of sending a zip file. That way you get properly credited. But I won't force any contribution that way.
And since I probably won't update this in a while, we won't have problems merging (unless this takes several months).

And again, thanks for your offer :) I'll wait for it!

**Comment by intrigus** [comment]
There is also https://github.com/ashleyj/aura which is based on robovm and aot compiles java.
It should work on linux and mac, but it uses GC so you probably can't use it.

**Comment by dsp-omen** [comment]
@intrigus Thanks! I knew about RoboVM and some of its forks, but what I really need is something without GC and ideally a single standard C++ file that I can compile with custom G++/clang compilers and glue with some native stuff easily.

@soywiz I have already done some work. Before starting the C++ target, I have unified code that was already duplicated in js/haxe target to avoid triplicating it. Now I'm going to start the target itself. Not sure how long it will take, but it shouldn't be too much since it is already prepared for multiple targets.

**Comment by dsp-omen** [comment]
I have a little surprise:

Actually I started the work two weeks ago familiarizing with the code, and last week started to code before reporting anything. So it is pretty advanced already, and probably I will send it to you this night or tomorrow.
It lacks some stuff, but I have fixed other things that I noticed were wrong.
I have some of the programs I wanted to get working working already and most of the tests are passing.
If you have time, it will require some refining work, and probably introduced some bugs. So if you don't have enough time, just tell me and I will do that in the following weeks with a slower pace.

**Comment by soywiz** [comment]
Wow! Is that true? It would be amazing :D
I'm eager to see it.
If it works fine, I will try to do a release this whole weekend.

Best regards

**Comment by soywiz** [comment]
To implement a GC for C++, I should look at this:
http://www.hpl.hp.com/hosted/linux/mail-archives/gc/2004-January/000281.html
`GC_gcj_malloc`

**Comment by intrigus** [comment]
This also looks interesting:
https://github.com/hsutter/gcpp

**Comment by soywiz** [comment]
Looks good.

My intention is to release 0.5.0 this weekend. It won't include more C++ work.
After that I will work on 0.6.0 with a new backend that will allow to remove all the artificial casts that prevents good performance on static targets like direct C++, D or Haxe.
Right now you can use haxe-cpp to generate C++ with garbage collector.
On 0.7.0 I will work on the relooper algorithm for targets without goto, working along the new backend. So in 0.7.0 all targets should work pretty fast.
If everything goes well this should be before the first year of public jtransc.
Between 0.8.0 and 1.0.0 I will check this and will decide which approach to use and try to learn a bit more about C++ to do this along with refining the API and cleaning up stuff.

**Comment by intrigus** [comment]
@dsp-omen hi, I first have to say thanks for the target you provided.
However I'd like to add support for more features, e.g. Threads and Atomics, but you said that you want to run the code on micro controllers, which may not support C++11.
That's why I need to know whether the compilers that you use for targeting your micro controllers also support C++11.

🎄 🎁

**Comment by soywiz** [comment]
@intrigus Last time I talk to dsp-omen via hangouts he told me that he was pretty busy and going to be offline for a while. So in the case he doesn't reply. I think that there is already a dependency on C++0x (so probably he have a toolchain supporting it):
https://github.com/jtransc/jtransc/blob/master/jtransc-gen-cpp/src/com/jtransc/gen/cpp/CppCompiler.kt#L24
As far as I noticed, probably because [std::shared_ptr](http://en.cppreference.com/w/cpp/memory/shared_ptr).
Also I asked him about GC, and he told me that he managed to totally avoid allocations for his use-case, so that I could feel free to change the shared pointer approach, that seemed to add some extra overhead to a GC approach.
Also, I don't think that Threads/Atomics it is a problem, since there is tree shaking that totally doesn't include code that is not being used. And if it was working for him, I think he was not referencing threads or atomics. Also in the worst case we could add something like `#if cpp11_available` or something like that.
So go ahead :) Also I have invited you to the jtransc organization, so you can merge your code yourself or directly work here. I have seen the JNI work and looks great :+1:

**Comment by soywiz** [comment]
@intrigus Didn't notice about this: "soywiz can you try MemberCollisionsTest on the cpp target?".
Sure I will test.

My main concern about IRC is that I cannot reply stuff, while you are offline or read it if it was not written while I'm online.
For that you can use slack, I'm usually connected here and I can read and reply being offline. You can join using this link: http://jtransc-slack.herokuapp.com/
And if you don't like slack, please write me/or hangout me to soywiz@gmail.com
